### ISSUE TITLE
Expose API for ethnographic annotations

### ISSUE THREAD

**Issue by albertocottica** [edgeryders/discourse#97]
In order to do social semantic network analysis, we need API access to both the content and the ethnographic annotations and codes generated via Open Ethnographer. The former is provided out of the box by Discourse's APIs. But the latter is not. Any solution you can think of, @damingo and @tanius ?

**Comment by tanius** [comment]
@damingo Sounds like adding a small JSON API in standard Ruby on Rails, as part of the `/administrator/` interface that you added for Open Ethnographer anyway. If so, how much time does that need?

**Comment by tanius** [comment]
@albertocottica This:

> we need API access to both the … ethnographic annotations and codes generated via Open Ethnographer

is an incomplete specification :wink: To be able to solve this, we need to know your API requirements in more detail. Of course the most basic API is "give me the attributes of annotation with ID …", and the same with codes. That is probably not enough, so tell us what else you need. Probably that means lists, such as:

* list of the IDs of all codes created by a given author
* list of the IDs of all annotations belonging to a given code
* list of the IDs of all annotations appearing in a given topic
* list of the IDs of all annotations appearing in a given post (= part of a topic)
* list of the IDs of all annotations created by a given author
* list of the IDs of all annotations in topics in a given category (and optional its sub-cateories)
* list of the IDs of all annotations in topics tagged with a given Discourse tag
* combinations of the above filters

**Comment by tanius** [comment]
@damingo Is the existing JSON interface of Annotator.js potentially enough as an implementation? What exactly can it do?

**Comment by albertocottica** [comment]
@tanius if we follow the Discourse example it is more dictionaries than lists.

The **main endpoint** is meant to retrieve the annotations associated to a given unit of content. This association allows us to induce a complete semantic social network. There are two ways to do it.

**Preferred solution: retrieve annotations by topic ID**. The endpoint would accept as argument the topic ID, and return annotation ID, snippet, code ID, _and post ID_; ideally code in text form too; and ethnographer ID. The disadvantage here is that the code needs to keep track of the post ID. But this, in my book, is more than compensated for by the dramatically reduced number of API calls (701 instead of 5,000 + in the OpenCare example).

Pseudocode:

````python
topics = fetch_topics_from tag('my-project) # a list of all topic IDs in the project
allPosts = []
allAnnotations = []
for top in topics:
  thisTopContent = fetch_posts_in_topic(top) # a list of dicts. Each dict encodes a post in the topic.
  allPosts.append(thisTopContent))  # the list gets appended to a global list of all posts in all topics
  thisTopAnnotations = fetch_annotations_in_topic(top) # a list of dicts. Each dict encodes one annotation in the topic
  allAnnotations.append(thisTopAnnotations) # the list gets appended to a global list of all annotations in all topics
full_info = integrate_annotations(allPosts, allAnnotations) # for each annotation, copies annotation fields into the dict representing the annotated post 
build_network (full_info)
````
**Alternative solution: retrieve annotations by post ID**. The post is the most granular unit of content in Discourse. The endpoint would accept as argument the post ID, and return annotation ID, snippet, code ID; ideally code in text form too; and ethnographer ID. The problem with this is that it the full induction of a semantic social network implies very many API calls. A project is denoted by a tag, so get topic IDs from the tag; then call each topic to get post IDs; finally, call each post ID to get the annotations. In OpenCare, this means calling one tag, 100 topics, and 4,350 posts.

The main endpoint is also used to quickly see what material has already been coded.

**Other endpoints**

1. retrieve all annotations by code ID. Of each annotation, display as above (if possible with the topic ID and post ID). Used in "quotation manager" type functions ("show me all the occurrences of `autonomy`" or "show me all the co-occurrences of `system_failure`and `legality`. These are used in graph reduction techniques. 
1. retrieve all annotations by collection. Of each annotation, . Used in displaying collections (= projects), make word clouds etc.
1. retrieve all annotations by author. Of each annotation, display as above. Used in displaying ethnographers' profiles: word cloud reflects interests etc.

Makes sense?

**Comment by tanius** [comment]
Makes sense.

**About the main endpoint proposals:** I propose we implement both main endpoints as "retrieve annotations by topic ID" will be internally implemented as multiple "retrieve annotations by post ID" anyway, so it's not much more effort but offers more flexibility (however that may or may not be used).

**About the API endpoint "retrieve all annotations by collection":** You say it's "used in displaying collections (= projects)". But these are two different things in my head. Collections are a not-yet-implemented feature in Open Ethnographer that allows grouping ethnographic codes (tags) together. Projects are groups of Discourse topics, now implemented with tags like `ethno-opencare` according to [your instructions](https://edgeryders.eu/t/7052). For that latter meaning of Open Ethnographer projects, the regular Discourse API can be used to retrieve the contained tags, so no additional endpoint is needed. An endpoint for collections can only be added once we implemented collections. So for now, let's drop this endpoint "retrieve all annotations by collection".

So what's the required time of completion for this feature?

(Assigning to @damingo already and marking as "important". You can already start implementing this.)

**Comment by albertocottica** [comment]
I stand corrected. A project is content-based: you start from one or more cats, maybe you add some tops from other cats, stick an `ethno-myproject`tag onto it and you are in business. At the origin of that, there almost always is an Edgeryders client-paid project. 

In my head, a collection is more an instrument for academic (secondary) research. It is (ethno) code-based: it works like this:

* A researcher (call her ego, in the sociology of networks tradition) starts browsing  codes. Note that in general these codes will have been authored by other researchers, not by ego herself. 
* She pulls in the codes she likes into a new collection.
* At this point, she automagically has a corpus. Because codes link to annotations, and annotations link to posts and topics, and a collection of posts an topics is a corpus. 
* She can then proceed to do several things. The easiest is to treat the corpus as an already coded corpus (which it is) and proceed to draw conclusions and write a paper or whatever on that basis. This is a "pure" case that I don't expect to really occur in reality. 
* Alternatively, she can proceed to recode the corpus, in part or in total. In this case, she sticks a new tag `ethno-mynewproject` onto all topics in the corpus; adds some annotations of her own; maybe adds the codes she uses to the collection; and _then_ does the analysis. The beauty of this is that ego has effectively enacted collaborative ethnography, even though she never even spoke to any of the ethnographers that came before her. In the end, she will write her paper on the basis of `n`annotations, of which `n_ego`will be her own, `n_alice`will be Alice's, `n_bob`will be Bob's etcetera. The collaborative ethnography documentation will contain advice as to how to credit these "implicit co-authors". 

Is it not beautiful? 😄 

**Time of completion**. 3-4 weeks OK? Amelia needs some time with GraphRyder to write her final report; and Jason needs some time to modify the GraphRyder code so it takes data from the Discourse API.

**Comment by tanius** [comment]
Pretty sure 3-4 weeks will be ok (though @damingo has to comment on that still).

As for collections, I think they need some more thought before we move to implementing them. It's going into the collaborative ethnography realm, so we need to finish the ["How to re-use codes of other authors?" discussion](https://edgeryders.eu/t/starting-ethnographic-research-activities/7257/2) first.

For the context of this issue of an initial API, I propose to redefine the "retrieve all annotations by collection" API endpoint as "retrieve all annotations by Discourse tag". That allows to work with the usual "corpus" marker tag, such as `ethno-opencare`.

So to summarize, API we're going to implement in the scope of this issue is:

* annotations by topic ID
* annotations by post ID
* annotations by code ID ("Open Ethnographer tag")
* annotations by Discourse tag
* annotations by author

**Comment by damingo** [comment]
Annotations endpoint is now accessible as https://edgeryders.eu/administration/annotator/annotations.json
To date the following GET attributes can be passed in to filter annotations:
* topic_id - Return annotations which belong to the given topic.
* post_id - Return annotations which belong to the given post.
* creator_id - Return annotations which were created by the given user.
* tag_id - Return annotations which tagged with the given tag.

Filter can be combined as needed. E.g. to return all annotations that belong to a certain topic and were created by a specific user:
https://edgeryders.eu/administration/annotator/annotations.json?topic_id=111&creator_id=222

By default not more than 100 annotations will be returned. This can be changed with the `per_page`attribute:
https://edgeryders.eu/administration/annotator/annotations.json?per_page=200
If there are more annotations available than the given `per_page` limit they can be accessed by setting the `page`attribute:
https://edgeryders.eu/administration/annotator/annotations.json?page=2
If no further annotations are available an empty array is returned.

@tanius How / from were to get the `code ID` ("Open Ethnographer tag") you mentioned?

**Comment by jason-vallet** [comment]
Hi,
Is it possible to open these routes and make their results publicly readable?

**Comment by albertocottica** [comment]
@damingo  I am wondering whether this can be solved with an API token, ideally reusing the Discourse API ones? It would be quite handy if access to annotations was restricted. Of course, we would give @jason-vallet the key. 😄

**Comment by tanius** [comment]
@damingo As discussed, the `tag_id` parameter should refer to a Discourse tag while the `code_ID` parameter that you asked about should refers to an Open Ethnographer tag (the tags you created the new table for). Please confirm here when it's implemented that way :)

Oh and please see if API access by Discourse API token can be implemented. Maybe it already works, as the `annotations.json` endpoint is accessible to staff members right now (moderators and admins), so perhaps also when staff members use an API key rather than a regular login? Has anyone tried @albertocottica?

**Comment by albertocottica** [comment]
@tanius it works for me. Navigating to  https://edgeryders.eu/administration/annotator/annotations.json when logged in with admin access results in a page styled in JSON:


```
[
  {
    id: 5579,
    version: "v1.0",
     text: null,
    quote: "comfortable for the patient.",
    uri: "/post/33751",
    created_at: "2017-05-22T19:13:10.000Z",
    updated_at: "2017-05-22T19:13:10.000Z",
    tag_id: 342,
    post_id: 33751,
    creator_id: 3323
  },
 ...
]   
```
(Why is the `text` field always evaluated to `null`?)

**Comment by tanius** [comment]
> Why is the `text` field always evaluated to `null`?

That will be because it's the free-text comment field of an annotation, which we do not use. The annotated text portion is contained in the `quote` field instead.

**Comment by damingo** [comment]
Filtering by discourse tag is now possible. To distinguish between annotator-tags and discourse-tags the `tag_id` attribute was renamed to `annotator_tag_id` and `discourse_tag` was added to allow filtering by discourse tag. I updated my comment above to reflect the change.

@albertocottica Is the current implementation sufficient to work with the API?

**Comment by tanius** [comment]
Ping @albertocottica :) Let's try to finish work on this issue in the next 4 days, if there's anything left to do at this point.

**Comment by albertocottica** [comment]
Not my department @tanius. What needs to happen now is for @jason-vallet to try and implement the GraphRyder harvester based on the new APIs. I suppose that they are accessible via API keys generated via the Discourse interface as usual, right? Jason, use of API keys is documented well enough in the Discourse documentation. I generated an API key for you: you will find it in your Eddgeryders profile.

**Comment by tanius** [comment]
Well then. Closing this as I conclude that the current implementation is _designed_ to be sufficient for GraphRyder. There may be issues / bugs in the implementation which we don't know yet. They should go into new issues when encountered.

API key access _should_ be available, but as I [mentioned above](#issuecomment-339449441), this is not confirmed yet.

**Comment by jason-vallet** [comment]
Hi, I have access to the annotations without trouble using the route <https://edgeryders.eu/administration/annotator/annotations.json> with my API and id but I am unable to get the annotator tags in a json format through <https://edgeryders.eu/administration/annotator/tags.json>. Is there a different route I am supposed to use?

The address  <https://edgeryders.eu/administration/annotator/tags> is working and returns all the necessary info but I would prefer to avoid parsing the html to get the tags and ids.

**Comment by tanius** [comment]
> I am unable to get the annotator tags in a json format

That is because no API for that exists yet. It's missing in my [summary of planned API functionality](#issuecomment-334971335) above and nobody complained :wink:

I'll create a new issue "Expose API for annotator tags" now to treat this shortcoming. Have also renamed this issue to reflect this "separation of concerns".

**Comment by damingo** [comment]
@jason-vallet As requested by @albertocottica in #111 ethnographer tags are now named "codes" to distinguish them from Discourse tags. The previous query parameter `annotator_tag_id` has been renamed to `code_id` to reflect the change.
